TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+                                    p.CodeElement.Column,
+                                    p.CodeElement.Column,
+                                    p.CodeElement.Line,
+                                    string.Format(Resource.BasicBlockGroupGoesBeyondTheLimit, ((BasicBlockForNode)block).Tag != null ? ((BasicBlockForNode)block).Tag.ToString() : "???", block.Index));
+                                Diagnostics.Add(d);
+                                //So in this case in order to not break the graph and to see the target branch that went out, add it as well....
+                                b.SuccessorEdges.Add(edge);
+                                continue;
+                            }
+                            else
+                            {//Don't add the continuation edge
+                                continue;
+                            }
+                        }
+                        var clonedBlock = this.CurrentProgramCfgBuilder.Cfg.AllBlocks[cloneBlockIndex];
+                        newEdge = this.CurrentProgramCfgBuilder.Cfg.SuccessorEdges.Count;
+                        this.CurrentProgramCfgBuilder.Cfg.SuccessorEdges.Add(clonedBlock);
+                        clonedEdgeIndexMap[edge] = newEdge;
+                    }
+                    b.SuccessorEdges.Add(newEdge);
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Resolve all pending PERFORM procedure.
+        /// </summary>
+        private void ResolvePendingPERFORMProcedures()
+        {
+            if (this.CurrentProgramCfgBuilder.PendingPERFORMProcedures != null)
+            {
+                foreach (var item in this.CurrentProgramCfgBuilder.PendingPERFORMProcedures)
+                {
+                    PerformProcedure p = item.Item1;
+                    BasicBlockForNodeGroup group = item.Item2;
+                    ResolvePendingPERFORMProcedure(p, group);
+                }
+                if (Mode == CfgMode.Extended)
+                {
+                    foreach (var item in this.CurrentProgramCfgBuilder.PendingPERFORMProcedures)
+                    {
+                        PerformProcedure p = item.Item1;
+                        BasicBlockForNodeGroup group = item.Item2;
+                        GraftBasicBlockGroup(group);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Compute the terminal blocks associated to Group
+        /// 
+        /// </summary>
+        /// <param name="group">The group to compute the terminal blocks</param>
+        private void ComputeBasicBlockGroupTerminalBlocks(BasicBlockForNodeGroup group)
+        {
+            if (group.Group.Count > 0)
+            {
+                LinkedListNode<BasicBlock<Node, D>> first = group.Group.First;
+                MultiBranchContext ctx = new MultiBranchContext(this.CurrentProgramCfgBuilder, null);
```

Comment:
We should not instantiate a context here. In my opinionThe `GetTerminalSuccessorEdges` method belongs to the `BasicBlockForNode` class. We'll have to add a reference to the graph in the block class though.

Reply:
It is possible to make the method GetTerminalSuccessorEdges a static, by adding a new parameter the target Cfg instance (ControlFlowGraph<Node, D> Cfg). But this method is really used in a MultiBranchContext logic associated to a Cfg construction not a Basic Block.